edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,134 @@
+extern crate nix;
+
+use std::{thread};
+use std::io::Error;
+use std::process::{Command, Stdio, Child};
+use std::sync::{Arc};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::time::Duration;
+use nix::unistd::Pid;
+use nix::sys::signal::{self, Signal};
+use signal_hook::{iterator::Signals, SIGTERM};
+use tracing::{info, error};
+use tracing_subscriber;
+
+fn main() -> Result<(), Error> {
+    init_logging();
+
+    let has_received_sigterm = match register_sigterm_listener() {
+        Ok(has_received_sigterm) => has_received_sigterm,
+        Err(e) => {
+            error!("Failed to register sigterm listener. Shutting down.");
+            return Err(e)
+        }
+    };
+
+    let mut edgehub = Command::new("dotnet")
+            .arg("/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll")
+            .stdout(Stdio::inherit())
```

Comment:
How would combined logs look like? Would that be readable?

Reply:
I thought that we would easily be able to filter based on the log headers, but that is not the case. Here is an example log of edgehub starting up:

```
Jul 06 21:47:12.324  INFO watchdog: Starting watchdog
Jul 06 21:47:12.402  INFO watchdog: Launched Edge Hub process with pid 8
Jul 06 21:47:12.403  INFO watchdog: Launched MQTT Broker process with pid 9
Jul 06 21:47:12.996  INFO mqttd: Loading state...
Jul 06 21:47:13.003  INFO mqtt_broker::persist: no state file found at /tmp/mqtt/state/state.dat.
Jul 06 21:47:13.004  INFO mqttd: state loaded.
Jul 06 21:47:13.004  INFO mqttd: Starting server...
Jul 06 21:47:13.004  INFO mqttd: Persisting state every 300s
Jul 06 21:47:13.004  INFO mqttd::snapshot::imp: Setup to persist state on USR1 signal
Jul 06 21:47:13.545  INFO mqtt_broker::transport: Loading identity from /tmp/mqtt/config/identity.pfx
Jul 06 21:47:13.620  INFO server{listener=0.0.0.0:1883}: mqtt_broker::server: Listening on address 0.0.0.0:1883
Jul 06 21:47:13.622  INFO server{listener=0.0.0.0:1883}:server{listener=0.0.0.0:8883}: mqtt_broker::server: Listening on address 0.0.0.0:8883
Jul 06 21:47:13.623  INFO server{listener=0.0.0.0:1883}:server{listener=0.0.0.0:8883}:server{listener=127.0.0.1:1882}: mqtt_broker::server: Listening on address 127.0.0.1:1882
2020-07-06 21:47:14.624 +00:00 Edge Hub Main()
<7> 2020-07-06 21:47:15.941 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Edged.WorkloadClient] - Making a Http call to unix:///var/run/iotedge/workload.sock to CreateServerCertificateAsync
<7> 2020-07-06 21:47:16.764 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Uds.HttpUdsMessageHandler] - Connecting socket /var/run/iotedge/workload.sock
<7> 2020-07-06 21:47:16.791 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Uds.HttpUdsMessageHandler] - Connected socket /var/run/iotedge/workload.sock
<7> 2020-07-06 21:47:16.803 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Uds.HttpUdsMessageHandler] - Sending request http://workload.sock/modules/%24edgeHub/genid/637244011215002617/certificate/server?api-version=2019-01-30
<7> 2020-07-06 21:47:16.927 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Uds.HttpUdsMessageHandler] - Response received Created
<7> 2020-07-06 21:47:16.972 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Edged.WorkloadClient] - Received a valid Http response from unix:///var/run/iotedge/workload.sock for CreateServerCertificateAsync
<7> 2020-07-06 21:47:17.359 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Edged.WorkloadClient] - Making a Http call to unix:///var/run/iotedge/workload.sock to TrustBundleAsync
<7> 2020-07-06 21:47:17.363 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Uds.HttpUdsMessageHandler] - Connecting socket /var/run/iotedge/workload.sock
<7> 2020-07-06 21:47:17.364 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Uds.HttpUdsMessageHandler] - Connected socket /var/run/iotedge/workload.sock
<7> 2020-07-06 21:47:17.364 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Uds.HttpUdsMessageHandler] - Sending request http://workload.sock/trust-bundle?api-version=2019-01-30
<7> 2020-07-06 21:47:17.365 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Uds.HttpUdsMessageHandler] - Response received OK
<7> 2020-07-06 21:47:17.367 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Util.Edged.WorkloadClient] - Received a valid Http response from unix:///var/run/iotedge/workload.sock for TrustBundleAsync
<6> 2020-07-06 21:47:17.378 +00:00 [INF] [EdgeHub] - Installing certificates [CN=iotedged workload ca:08/18/2020 16:01:01],[CN=Test Edge Device CA:08/18/2020 16:01:01],[CN=Test Edge Owner CA:08/18/2020 16:01:01] to Root
<6> 2020-07-06 21:47:17.417 +00:00 [INF] [EdgeHub] - Installing certificates [CN=Test Edge Owner CA:08/18/2020 16:01:01] to Root
<6> 2020-07-06 21:47:17.430 +00:00 [INF] [EdgeHub] - Enabling SSL protocols: Tls, Tls11, Tls12
<6> 2020-07-06 21:47:18.216 +00:00 [INF] [EdgeHub] - Experimental features configuration: {"Enabled":false,"DisableCloudSubscriptions":false,"DisableConnectivityCheck":false}
<6> 2020-07-06 21:47:19.841 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Service.Modules.RoutingModule] - Created persistent store at /tmp/edgeHub
<6> 2020-07-06 21:47:19.862 +00:00 [INF] [EdgeHub] - Initializing Edge Hub

```

I would like to find a way to prepend appropriate log headers to each line so that one can effectively filter by watchdog, broker, or edgehub logs. Something like [Edge Hub], [Broker], [Watchdog]. I'm not sure if this is possible from the watchdog or we will have to do it from the edgehub/broker themselves. I'll check. What do you think of this idea?